File: cpp/easy/FindTarget.cpp
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

bool findTarget(TreeNode *root, int k) {
  unordered_set<int> set;
  stack<TreeNode *> stack;
  while (root || !stack.empty()) {
    if (root) {
      if (set.count(root->val)) {
        return true;
      }
      set.insert(k - root->val);
      stack.push(root);
      root = root->left;
    } else {
      root = stack.top();
      stack.pop();
      root = root->right;
    }
  }
  return false;
}

File: go/easy/findTarget.go
package main

var (
	m  map[int]int
	rs bool = false
)

func recur(root *TreeNode, k int) {
	if root == nil {
		return
	}
	if _, exist := m[root.Val]; exist {
		rs = true
		return
	}
	m[k-root.Val] = 1
	recur(root.Left, k)
	recur(root.Right, k)
}

func findTarget(root *TreeNode, k int) bool {
	rs = false
	m = make(map[int]int)
	recur(root, k)
	return rs
}
